### PR TITLE
Lewis

### PR DESCRIPTION
**Pull request by lew101**

### CODE REVIEW

File: minitask3/Lewis/follow.py [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+        self.plot_subscriber = rospy.Subscriber("odom", Odometry, self.plot_trajectory)
+
+        # For obstacle avoidance
+        self.laser_sub = rospy.Subscriber('/scan', LaserScan, self.scan_callback)
+        self.obstacle = True
+        self.direction = 0
+        self.direction_set = False
+
+        # For odometry and position tracking
+        self.curr_pose = {}
```
Comment (bhenaghan) [inline minitask3/Lewis/follow.py]: You can initialise all of these in one line with `self.curr_pose = {'theta': 0.0...` etc.

File: minitask3/Lewis/follow.py [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+    def image_callback(self, data):
+        '''
+        Handles the camera data.
+        '''
+        try:
+            cv_image = self.bridge.imgmsg_to_cv2(data,'bgr8')
+        except CvBridgeError as error:  # catch conversion errors
+            print(error)
+
+        cv_image_resized = cv2.resize(cv_image, (cv_image.shape[1]/4,cv_image.shape[0]/4))
```
Comment (bhenaghan) [inline minitask3/Lewis/follow.py]: This line fill fail if line `102` fails and the except is triggered, as cv_image will be `None`

Comment (bhenaghan) [reply minitask3/Lewis/follow.py]: You're also missing spaces around the `cv_image.shape[1]/4`, I would probably split that tuple into two lines.

File: minitask3/Lewis/follow.py [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+        '''
+        contours, _ = cv2.findContours(mask.copy(),
+                                        cv2.RETR_TREE,
+                                        cv2.CHAIN_APPROX_SIMPLE,
+                                        offset=(0, 0))  # Finds the outlines of each object
+        centroids = []
+        for contour in contours:  # Creates an array of the centres of each object
+            moments = cv2.moments(contour)
+            if moments["m00"] != 0:
+                center_x = int(moments['m10']/moments['m00'])
```
Comment (bhenaghan) [inline minitask3/Lewis/follow.py]: Missing spaces

File: minitask3/Lewis/follow.py [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+
+        return centroids
+
+    def get_mask(self, image):
+        '''
+        Returns an image mask based on the colour green
+        '''
+        image_hsv = cv2.cvtColor(image, cv2.COLOR_BGR2HSV)
+        low_green = np.array([0, 100, 0])
+        high_green = np.array([100, 255, 255])
```
Comment (bhenaghan) [inline minitask3/Lewis/follow.py]: might be nice to extract these as constants

**Review by BenH14 — APPROVED** [review]
LGTM! Looks really nice, I've only found a couple of little things that could be changed.

**Review by akhmaddias — APPROVED** [review]
LGTM!

**Comment by lew101** [comment]
I've made those changes!

I also realised that the scan method wasn't needed anymore after those constants were added so I have removed it

**Pull request merged**